### PR TITLE
[v2] Update protocol to use new UnitExpectedConfig.

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

This updates the github.com/elastic-agent/elastic-agent-client to use and send the new `UnitExpectedConfig` to over the v2 protocol. The `ExpectedConfig` function takes a `map[string]interface{}` and turns it into the `UnitExpectedConfig` with `Source` set.

Because it is possible that decoding the YAML into a `UnitExpectedConfig` a unit can be marked failed from the Elastic Agent side and it will never be sent to the running component. It will stay marked as Failed until it is corrected in the configuration. Once corrected it will then be sent to the component.

## Why is it important?

Path to v2 architecture.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

### CODE REVIEW

**Comment by blakerouse** [comment]
@michalpristas @narph This is now ready for a full review.

File: pkg/component/component.go [hunk 1]
```diff
@@ -255,10 +295,24 @@ func toIntermediate(policy map[string]interface{}) (map[string]outputI, error) {
...
+		logLevel := defaultUnitLogLevel
+		if logLevelRaw, ok := input[logLevelKey]; ok {
+			logLevelStr, ok := logLevelRaw.(string)
+			if !ok {
+				return nil, fmt.Errorf("invalid 'inputs.%d.log_level', expected a string not a %T", idx, logLevelRaw)
+			}
+			var err error
+			logLevel, err = stringToLogLevel(logLevelStr)
+			if err != nil {
+				return nil, fmt.Errorf("invalid 'inputs.%d.log_level', %w", idx, err)
```
Comment (ph) [inline pkg/component/component.go:307]: Nit: Maybe we could tell the supported log level?

File: pkg/component/runtime/manager_test.go [hunk 2]
```diff
@@ -253,6 +255,323 @@ LOOP:
...
+		case err := <-subErrCh:
+			require.NoError(t, err)
+			break LOOP
+		}
+	}
+
+	subCancel()
+	cancel()
+
+	err = <-errCh
```
Comment (aleksmaus) [inline pkg/component/runtime/manager_test.go:421]: nit: maybe use errgroup here?

File: pkg/component/runtime/manager_test.go [hunk 3]
```diff
@@ -253,6 +255,323 @@ LOOP:
...
+		case err := <-errCh:
+			require.NoError(t, err)
+		case err := <-subErrCh:
+			require.NoError(t, err)
+			break LOOP
+		}
+	}
+
+	subCancel()
+	cancel()
```
Comment (aleksmaus) [inline pkg/component/runtime/manager_test.go:569]: errgroup? any refactoring, seems some repetitive code here

**Review by aleksmaus — APPROVED** [review]
LGTM
left few comments/nits

**Review by ph — APPROVED** [review]
LGTM, +1 on the errgroup suggestion from @aleksmaus

File: pkg/component/component.go [hunk 4]
```diff
@@ -255,10 +295,24 @@ func toIntermediate(policy map[string]interface{}) (map[string]outputI, error) {
...
+			}
+			delete(input, logLevelKey)
+		}
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/component/component.go:310]: Refactor it into a function.

File: pkg/component/config.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	result := &proto.UnitExpectedConfig{}
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		ZeroFields:           true,
+		WeaklyTypedInput:     true,
+		TagName:              "json",
+		IgnoreUntaggedFields: true,
+		Result:               result,
+		MatchName: func(mapKey, fieldName string) bool {
+			if fieldName == sourceFieldName {
+				// never match for 'source' field that is set manually after
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/component/config.go:46]: Updated.

**Review by ph — APPROVED** [review]
changes lgtm, @fearful-symmetry could you take a look at this?

**Review by fearful-symmetry — APPROVED** [review]

**Pull request merged**